### PR TITLE
Support http publisher

### PR DESCRIPTION
**Pull request by MarcoPolo**
## Context

go-legs supports syncing over http. But the index provider hasn't been hooked up to start the http server.

## Changes

This PR adds a config and logic to enable the http index provider via go-legs. It also includes a test in `e2e_retrieve_test.go` that syncs over http.

### CODE REVIEW

File: engine/engine.go [hunk 1]
```diff
@@ -212,7 +231,19 @@ func (e *Engine) Publish(ctx context.Context, adv schema.Advertisement) (cid.Cid
 
 	log.Infow("Publishing advertisement in pubsub channel", "cid", c.String())
 	// Use legPublisher to publish the advertisement.
-	return c, e.publisher.UpdateRoot(ctx, c)
+	err = e.publisher.UpdateRoot(ctx, c)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	if e.httpPublisher != nil {
```
Comment (willscott) [inline engine/engine.go:239]: can we have the publisher be one or the other, rather than both a graphsync and http publisher?

File: config/httpserver.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	httpMultiaddr, _ := multiaddr.NewMultiaddr("/http")
+	maddr = maddr.Decapsulate(httpMultiaddr)
+
+	netAddr, err := manet.ToNetAddr(maddr)
+	if err != nil {
+		return "", err
+	}
+	return netAddr.String(), nil
+}
```
Comment (gammazero) [inline config/httpserver.go:34]: This seems like the wrong package for this file.  The config package should only be concerned about settings read from the config file. 

I think that `ListenMultiaddr` [here](https://github.com/filecoin-project/index-provider/blob/main/config/providerserver.go#L5) should used.  If it is an HTTP address then start the http publisher, otherwise start the datatransfer publisher.  The rest should go under `server/`.

**Review by gammazero — APPROVED** [review]

File: engine/engine.go (hunk 1 above)
Comment (MarcoPolo) [reply engine/engine.go]: done

File: config/httpserver.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package config
+
+import (
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+type HttpServer struct {
+	Enabled         bool
```
Comment (masih) [inline config/httpserver.go:9]: Please document behaviour of this flag; enabling HTTP publisher seems to disable dtsync publisher?

File: engine/engine.go [hunk 4]
```diff
@@ -161,11 +165,23 @@ func (e *Engine) Start(ctx context.Context) error {
 
 	e.cache = dsCache
 
-	e.publisher, err = dtsync.NewPublisherFromExisting(e.dataTransfer, e.host, e.pubSubTopic, e.lsys)
+	if e.httpPublisherCfg.Enabled {
```
Comment (masih) [inline engine/engine.go:168]: So means Provider would start either a HTTP publisher or `dtsync.Publusher`? 

I think we need to document this.

File: engine/engine.go [hunk 5]
```diff
@@ -258,11 +278,13 @@ func (e *Engine) Shutdown() error {
 	err := e.publisher.Close()
 	if err != nil {
 		err = fmt.Errorf("error closing leg publisher: %w", err)
+		return err
```
Comment (masih) [inline engine/engine.go:281]: This means `e.cache` will not close if we fail to close the publisher. 

I am curious why this change?

Comment (MarcoPolo) [reply engine/engine.go:281]: ah yes, I mean to undo this. Thanks for the catch!

File: engine/engine.go (hunk 4 above)
Comment (MarcoPolo) [reply engine/engine.go]: Changed in 0a8a8bb1697d77109f30f20627fc034e77da98f4 with an enum field in the ingest cfg.

File: config/httpserver.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package config
+
+import (
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+type HttpServer struct {
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply config/httpserver.go]: Done 336fa9deca0b78cfcdab22f7b75bc7f6b0d16772

File: config/httpserver.go (hunk 3 above)
Comment (MarcoPolo) [reply config/httpserver.go]: Changed via enum: 0a8a8bb1697d77109f30f20627fc034e77da98f4

**Comment by MarcoPolo** [comment]
Great feedback @masih! Updated with your suggestions

**Review by gammazero — APPROVED** [review]
LGTM
